### ISSUE TITLE
DB::DeleteFile vs DB::CompactRange race

### ISSUE THREAD

**Issue by ttyusupov** [facebook/rocksdb#8491]
> Note: Please use Issues only for bug reports. For questions, discussions, feature requests, etc. post to dev group: https://groups.google.com/forum/#!forum/rocksdb or https://www.facebook.com/groups/rocksdb.dev

### Expected behavior

DeleteFile fails, Manual compaction completed successfully, after that writes into RocksDB and flush succeeds.

### Actual behavior

Manual compaction fails with "Corruption: Compaction input files inconsistent", subsequent writes and flush fail with the same error:
```
./deletefile_test --gtest_filter=DeleteFileTest.DeleteWithManualCompaction
Note: Google Test filter = DeleteFileTest.DeleteWithManualCompaction
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from DeleteFileTest
[ RUN      ] DeleteFileTest.DeleteWithManualCompaction
db/deletefile_test.cc:583: Failure
manual_compaction_status
Corruption: Compaction input files inconsistent
db/deletefile_test.cc:62: Failure
db_->Put(options, key, value)
Corruption: Compaction input files inconsistent
db/deletefile_test.cc:586: Failure
dbi->TEST_FlushMemTable( true)
Corruption: Compaction input files inconsistent
[  FAILED  ] DeleteFileTest.DeleteWithManualCompaction (1517 ms)
```

### Steps to reproduce the behavior

1) `git checkout v6.20.3`
2) `git apply` the following patch:
```
diff --git a/db/deletefile_test.cc b/db/deletefile_test.cc
index 38bc49602..09ee630bd 100644
--- a/db/deletefile_test.cc
+++ b/db/deletefile_test.cc
@@ -551,6 +551,43 @@ TEST_F(DeleteFileTest, DeleteNonDefaultColumnFamily) {
   }
 }

+TEST_F(DeleteFileTest, DeleteWithManualCompaction) {
+  constexpr auto kNumSst = 3;
+  constexpr auto kNumKeysPerSst = 10000;
+
+  DBImpl* dbi = reinterpret_cast<DBImpl*>(db_);
+  for (auto num = 0; num < kNumSst; num++) {
+    AddKeys(kNumKeysPerSst, 0);
+    ASSERT_OK(dbi->TEST_FlushMemTable(/* wait = */ true));
+  }
+
+  std::vector<LiveFileMetaData> metadata;
+  db_->GetLiveFilesMetaData(&metadata);
+
+  std::atomic<bool> manual_compaction_done{false};
+  Status manual_compaction_status;
+
+  std::thread manual_compaction([this, &manual_compaction_done, &manual_compaction_status] {
+    manual_compaction_status = db_->CompactRange(
+        rocksdb::CompactRangeOptions(), /* begin = */ nullptr, /* end = */ nullptr);
+    manual_compaction_done = true;
+  });
+
+  while (!manual_compaction_done) {
+    for (auto& file_meta : metadata) {
+      db_->DeleteFile(file_meta.name);
+    }
+  }
+
+  manual_compaction.join();
+  EXPECT_OK(manual_compaction_status);
+
+  AddKeys(10, 0);
+  ASSERT_OK(dbi->TEST_FlushMemTable(/* wait = */ true));
+
+  Close();
+}
+
 }  // namespace ROCKSDB_NAMESPACE

 #ifdef ROCKSDB_UNITTESTS_WITH_CUSTOM_OBJECTS_FROM_STATIC_LIBS
```
3) `./deletefile_test --gtest_filter=DeleteFileTest.DeleteWithManualCompaction`

**Comment by ajkr** [comment]
I was hoping to remove `DeleteFile()` (see #7337). Do you mind explaining the use case?

**Comment by ttyusupov** [comment]
@ajkr , we use it in https://github.com/yugabyte/yugabyte-db to delete files we know for sure we no longer need to avoid wasting CPU time on compactions when it is not needed.

**Comment by ajkr** [comment]
Understood the use case, thanks. I wonder if we can support a better way of filtering files out based on properties; it seems you could delete even more files containing only obsolete intents if we had it (since currently `DeleteFile()` only works for the bottom level)?

**Comment by ajkr** [comment]
For example we could expose a callback like this for use in compaction: https://github.com/facebook/rocksdb/blob/870033291ab00447dfebdba29361df4ca13799cc/include/rocksdb/options.h#L1458-L1464.

Then the files you were deleting before would be picked for compaction (perhaps through the marking mechanism), but compaction execution will skip reading from them so their data won't be output anywhere.

**Comment by ttyusupov** [comment]
> it seems you could delete even more files containing only obsolete intents if we had it (since currently DeleteFile() only works for the bottom level)?

YugabyteDB currently uses only one level.

**Comment by ttyusupov** [comment]
> Then the files you were deleting before would be picked for compaction (perhaps through the marking mechanism), but compaction execution will skip reading from them so their data won't be output anywhere.

What if we want to delete some obsolete files, but we don't want the rest to be compacted to save CPU resources?

**Comment by ttyusupov** [comment]
>> Then the files you were deleting before would be picked for compaction (perhaps through the marking mechanism), but compaction execution will skip reading from them so their data won't be output anywhere.

1)
> What if we want to delete some obsolete files, but we don't want the rest to be compacted to save CPU resources?

2) Another scenario is that we need to delete some obsolete files before they got included in the compaction (not all compactions are full).

**Comment by ajkr** [comment]
> > Then the files you were deleting before would be picked for compaction (perhaps through the marking mechanism), but compaction execution will skip reading from them so their data won't be output anywhere.
> 
> What if we want to delete some obsolete files, but we don't want the rest to be compacted to save CPU resources?

Yes, this would currently be a problem. I will see if there's a reasonable way to trigger a single-file compaction with a table_filter that could drop that file, which would essentially delete it without reading it (but still technically going through the "compaction" process to unify code paths).
